### PR TITLE
Backward Paths for Tomography

### PR DESCRIPTION
**Pull request by johertrich**
PR regarding the discussion in #511 

It implements the following fixes/changes to the tomography operator:
- The backward path for the tomography operator is now set to the adjoint and vice versa. 
- There is an option `adjoint_via_backprop` to use the `deepinv.physics.adjoint_function` instead of `iradon` as adjoint and backward path.
- If the option `circle=True` is selected, the forward operator sets all values outside of the circle to zero before applying the radon transform (otherwise operator and the adjoint via iradon do not match and the adjoint via backprop effects pixels outside the circle).
- I added an option `fbp_interpolate_boundary` (default `False`), which enhances the FBP a bit. Currently, due to the padding with `circle=False`, there often appear white artifacts at the boundary (I guess these are streaking artifacts due to the artificial edge introduced by the padding). If `fbp_interpolate_boundary=True`, the outer two pixels are cut off and replaced by an interpolation by based on the remaining image.

I also fixed a bug and added an option in the PSNR implementation: 
- if `max_pixel is None` and a batch of images is given, it takes the maximum pixel of all ground truth images in the batch. This is completely counterintuitive to me, since then evaluating the PSNR on a batch is not the same as evaluating the PSNR on any image and taking the average. I fixed the batching for the `max_pixel=None` option.
- I also added a `min_pixel` option analogously to `max_pixel` such that it can really choose an adaptive range.

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -14,6 +14,7 @@ Changed
 ^^^^^^^
 - Make autograd use the base linear operator for `deepinv.physics.adjoint_function` (:gh:`519` by `Jérémy Scanvic`_)
 - Parallelize the test suite making it 15% faster (:gh:`522` by `Jérémy Scanvic`_)
+- Adjust backward paths for tomography (:gh:`535` by `Johannes Hertrich`_)
```
Comment (Andrewwango) [inline CHANGELOG.rst:19]: Please can you add your PSNR `max_pixel` fix to changelog?

File: deepinv/loss/metric/distortion.py [hunk 2]
```diff
@@ -210,21 +210,32 @@ class PSNR(Metric):
...
+    def __init__(self, max_pixel=1, min_pixel=0, **kwargs):
         super().__init__(**kwargs)
         self.max_pixel = max_pixel
+        self.min_pixel = min_pixel
         self.lower_better = False
 
     def metric(self, x_net, x, *args, **kwargs):
-        max_pixel = self.max_pixel if self.max_pixel is not None else x.max()
-        return cal_psnr(x_net, x, max_pixel=max_pixel)
+        max_pixel = (
```
Comment (Andrewwango) [inline deepinv/loss/metric/distortion.py:263]: Great fix, thanks. If you have time please can you also add this to `SSIM`?

File: deepinv/physics/tomography.py [hunk 3]
```diff
@@ -32,6 +33,11 @@ class Tomography(LinearPhysics):
     :param bool circle: If ``True`` both forward and backward projection will be restricted to pixels inside a circle
         inscribed in the square image.
     :param bool parallel_computation: if True, all projections are performed in parallel. Requires more memory but is faster on GPUs.
+    :param bool adjoint_via_backprop: if True, the adjoint will be computed via ``deepinv.physics.adjoint_function``. Otherwise the inverse Radon transform is used.
```
Comment (Andrewwango) [inline deepinv/physics/tomography.py]:
```suggestion
    :param bool adjoint_via_backprop: if `True`, the adjoint will be computed via :func:`deepinv.physics.adjoint_function`. Otherwise the inverse Radon transform is used.
```

File: deepinv/physics/tomography.py [hunk 4]
```diff
@@ -32,6 +33,11 @@ class Tomography(LinearPhysics):
...
+        The inverse Radon transform is computationally cheaper (particularly in memory), but has a small adjoint mismatch.
+        The backprop adjoint is the exact adjoint, but might break random seeds since it backpropagates through ``torch.nn.functional.grid_sample``, see the note `here <https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html>`_,
```
Comment (Andrewwango) [inline deepinv/physics/tomography.py]:
```suggestion
        The backprop adjoint is the exact adjoint, but might break random seeds since it backpropagates through :func:`torch.nn.functional.grid_sample`, see the note `here <https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html>`_.
```

File: deepinv/physics/functional/radon.py [hunk 5]
```diff
@@ -423,3 +438,34 @@ def _create_grids(self, angles, grid_size, circle):
                 torch.cat((X.unsqueeze(-1), Y.unsqueeze(-1)), dim=-1).unsqueeze(0)
             )
         return torch.stack(all_grids)
+
+
+# autograd wrapper
+class ApplyRadon(torch.autograd.Function):
```
Comment (Andrewwango) [inline deepinv/physics/functional/radon.py:491]: Please can you add class docstring and methods docstrings and types?

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Small formatting issues thanks!

**Comment by johertrich** [comment]
Hi Andrew, thanks for your comments. I implemented the changes.

File: deepinv/tests/test_physics.py [hunk 6]
```diff
@@ -1054,27 +1054,32 @@ def test_tomography(device):
...
-                    r = (
-                        physics.A_adjoint(physics.A(x))
-                        * PI.item()
-                        / (2 * len(physics.radon.theta))
-                    )
-                    y = physics.A(r)
-                    error = (physics.A_dagger(y) - r).flatten().mean().abs()
-                    epsilon = 0.2 if device == "cpu" else 0.3  # Relax a bit of GPU
-                    assert error < epsilon
+                    for adjoint_via_backprop in [True, False]:
```
Comment (Andrewwango) [inline deepinv/tests/test_physics.py]: One further formatting comment, please can you pull from #541 and adapt your contributions to the test similarly? i.e. use pytest parametrize rather than for loop. Thanks!

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Hi @johertrich , one last formatting comment for now. Leaving it for others to review the actual technical contribution of the PR :P

File: deepinv/physics/functional/radon.py [hunk 7]
```diff
@@ -221,6 +221,21 @@ def forward(self, x):
             pad_width = (pad_before, pad - pad_before)
             x = F.pad(x, (pad_width[0], pad_width[1], pad_width[0], pad_width[1]))
 
+        if self.circle:
+            yax = (
+                2
+                * (
+                    torch.arange(W, dtype=torch.float, device=x.device)[None, :].repeat(
```
Comment (mh-nguyen712) [inline deepinv/physics/functional/radon.py]: Minor comment: as you don't change value of yax, using `expand` instead of `repeat` will avoid copying data

File: deepinv/physics/functional/radon.py [hunk 8]
```diff
@@ -221,6 +221,21 @@ def forward(self, x):
...
+                        W, 1
+                    )[None, None, :, :]
```
Comment (mh-nguyen712) [inline deepinv/physics/functional/radon.py]:
```suggestion
                    torch.arange(W, dtype=torch.float, device=x.device)[None, :].expand(
                        W, -1
                    )[None, None, :, :]
```

File: deepinv/physics/tomography.py [hunk 9]
```diff
@@ -139,27 +153,41 @@ def __init__(
...
+            y = self.iradon.filter(y)
+            output = (
+                ApplyRadon.apply(y, self.radon, self.iradon, True)
+                * torch.pi
+                / (2 * len(self.iradon.theta))
+            )
             if self.normalize:
                 output = output * output.shape[-1]
+        if self.fbp_interpolate_boundary:
+            output = output[:, :, 2:-2, 2:-2]
```
Comment (mh-nguyen712) [inline deepinv/physics/tomography.py:218]: Sorry I'm not very familiar with CT, but is there any reason that the padding is 2? Or it's a value that make sense.

File: deepinv/tests/test_physics.py [hunk 10]
```diff
@@ -1043,38 +1043,46 @@ def test_reset_noise(device):
...
-                        physics.A_adjoint(physics.A(x))
-                        * PI.item()
-                        / (2 * len(physics.radon.theta))
-                    )
-                    y = physics.A(r)
-                    error = (physics.A_dagger(y) - r).flatten().mean().abs()
-                    epsilon = 0.2 if device == "cpu" else 0.3  # Relax a bit of GPU
-                    assert error < epsilon
+    x = torch.randn(imsize, device=device).unsqueeze(0)
+    r = physics.A_adjoint(physics.A(x)) * PI.item() / (2 * len(physics.radon.theta))
```
Comment (mh-nguyen712) [inline deepinv/tests/test_physics.py]:
```suggestion
    r = physics.A_adjoint(physics.A(x)) * torch.pi / (2 * len(physics.radon.theta))
```

File: deepinv/tests/test_physics.py [hunk 11]
```diff
@@ -1043,38 +1043,46 @@ def test_reset_noise(device):
...
+    adjoint_via_backprop,
+    fbp_interpolate_boundary,
+    device,
+):
     r"""
     Tests tomography operator which does not have a numerically precise adjoint.
 
     :param device: (torch.device) cpu or cuda:x
     """
     PI = 4 * torch.ones(1).atan()
```
Comment (mh-nguyen712) [inline deepinv/tests/test_physics.py]:
```suggestion

```

**Review by mh-nguyen712 — COMMENTED** [review]
Many thanks for your PR, look good to me except some minor comments

File: deepinv/physics/tomography.py (hunk 9 above)
Comment (johertrich) [reply deepinv/physics/tomography.py:218]: If circle=False, the input image will be padded to size sqrt(2)*input_size. This introduces an artificial edge in the image. When using the FBP this edge leads to a streaking artifact only effecting very few boundary pixels. This option implements a small trick by first cutting off two pixels at the boundary and then reinterpolate them (this is quite heuristic). I would keep it optional with default False, but it can help in some cases (and often reduces the error $$||A^T A \mathrm{FBP}(x) - A(x)|| $$ signficantly, i.e., it is closer to Moore-Penrose inverse than the FBP). In my experiences 2 pixels are enough, but if you prefer, I can also set is as an input argument.

Comment (mh-nguyen712) [reply deepinv/physics/tomography.py:218]: Thanks for your clarification. If 2 pixels are enough, I think we can leave as it for simplicity.

File: deepinv/physics/tomography.py [hunk 12]
```diff
@@ -31,7 +32,13 @@ class Tomography(LinearPhysics):
...
     :param bool circle: If ``True`` both forward and backward projection will be restricted to pixels inside a circle
         inscribed in the square image.
-    :param bool parallel_computation: if True, all projections are performed in parallel. Requires more memory but is faster on GPUs.
+    :param bool parallel_computation: if ``True``, all projections are performed in parallel. Requires more memory but is faster on GPUs.
+    :param bool adjoint_via_backprop: if ``True``, the adjoint will be computed via :func:`deepinv.physics.adjoint_function`. Otherwise the inverse Radon transform is used.
+        The inverse Radon transform is computationally cheaper (particularly in memory), but has a small adjoint mismatch.
+        The backprop adjoint is the exact adjoint, but might break random seeds since it backpropagates through :func:`torch.nn.functional.grid_sample`, see the note `here <https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html>`_.
+    :param bool fbp_interpolate_boundary: the FBP (``A_dagger``) usually contains streaking artifacts on the boundary due to padding. For ``fbp_interpolate_boundary=True``
```
Comment (tachella) [inline deepinv/physics/tomography.py]:
```suggestion
    :param bool fbp_interpolate_boundary: the :func:`filtered back-projection <deepinv.physics.Tomography.A_dagger>` usually contains streaking artifacts on the boundary due to padding. For ``fbp_interpolate_boundary=True``
```

File: deepinv/physics/tomography.py [hunk 13]
```diff
@@ -139,27 +153,41 @@ def __init__(
...
-        output = self.radon(x)
+        if self.fan_beam or self.adjoint_via_backprop:
+            output = self.radon(x)
+        else:
+            output = ApplyRadon.apply(x, self.radon, self.iradon, False)
         if self.normalize:
             output = output / x.shape[-1]
         return output
 
     def A_dagger(self, y, **kwargs):
```
Comment (tachella) [inline deepinv/physics/tomography.py:182]: It would be nice to add a docstring here (I don't know why it wasn't there)
```suggestion
    def A_dagger(self, y, **kwargs):
    r"""
    Computes the filtered back-projection (FBP) of the measurements.
    
    .. warning::

        The filtered back-projection algorithm is not the exact linear pseudo-inverse of the Radon transform, but it is a good approximation that is robust to noise.
        
        .. tip::
        
             By default, the FBP reconstruction can display artifacts at the borders. Set ``fbp_interpolate_boundary=True`` to remove them with padding.
        
        
        :param torch.Tensor y: measurements
        :return torch.Tensor: noisy measurements
    """
```

File: deepinv/physics/tomography.py [hunk 14]
```diff
@@ -31,7 +32,13 @@ class Tomography(LinearPhysics):
     :param int img_width: width/height of the square image input.
     :param bool circle: If ``True`` both forward and backward projection will be restricted to pixels inside a circle
         inscribed in the square image.
-    :param bool parallel_computation: if True, all projections are performed in parallel. Requires more memory but is faster on GPUs.
+    :param bool parallel_computation: if ``True``, all projections are performed in parallel. Requires more memory but is faster on GPUs.
+    :param bool adjoint_via_backprop: if ``True``, the adjoint will be computed via :func:`deepinv.physics.adjoint_function`. Otherwise the inverse Radon transform is used.
```
Comment (tachella) [inline deepinv/physics/tomography.py:51]: It would be nice to mention this option in the warning above, that is move 


```
    .. warning::

        The adjoint operator has small numerical errors due to interpolation.
```

to 

```
    .. warning::

        The default adjoint operator has small numerical errors due to interpolation. Set ``adjoint_via_backprop=True`` if you want to use the exact adjoint (computed via autograd).
```

File: deepinv/physics/tomography.py [hunk 15]
```diff
@@ -139,27 +153,41 @@ def __init__(
...
+                / (2 * len(self.iradon.theta))
+            )
             if self.normalize:
                 output = output * output.shape[-1]
+        if self.fbp_interpolate_boundary:
+            output = output[:, :, 2:-2, 2:-2]
+            output = torch.nn.functional.pad(output, (2, 2, 2, 2), mode="replicate")
         return output
 
     def A_adjoint(self, y, **kwargs):
```
Comment (tachella) [inline deepinv/physics/tomography.py:222]:
```suggestion
    def A_adjoint(self, y, **kwargs):
    r"""
        Computes adjoint of the tomography operator.
    
        .. warning::

        The default adjoint operator has small numerical errors due to interpolation. Set ``adjoint_via_backprop=True`` if you want to use the exact adjoint (computed via autograd).
        
        :param torch.Tensor y: measurements
        :return torch.Tensor: noisy measurements
        """
```

File: deepinv/tests/test_physics.py [hunk 16]
```diff
@@ -1061,6 +1071,8 @@ def test_tomography(normalize, parallel_computation, fan_beam, circle, device):
         circle=circle,
         fan_beam=fan_beam,
         normalize=normalize,
+        adjoint_via_backprop=adjoint_via_backprop,
```
Comment (tachella) [inline deepinv/tests/test_physics.py:1079]: Now that the tomography operator is working well with the `adjoint_via_backprop` option, I think it would be nice to to test_adjointness in this function when `adjoint_via_backprop=True`, something like

```
if adjoint_via_backprop=True:
     assert physics.adjointness_test(x).abs() < 0.001

```

**Review by tachella — APPROVED** [review]

File: deepinv/physics/functional/radon.py [hunk 17]
```diff
@@ -470,3 +485,45 @@ def _create_grids(self, angles, grid_size, circle):
...
+
+
+# autograd wrapper
+class ApplyRadon(torch.autograd.Function):
+    r"""
+    (Static) Autograd Wrapper for the Radon transform and its adjoint.
+
+    It can be called by ``ApplyRadon.apply(x, radon, iradon, adjoint)`` and with inputs specified as follows:
+
+    :param torch.Tensor x:
```
Comment (Andrewwango) [inline deepinv/physics/functional/radon.py]: Description here

File: deepinv/physics/tomography.py [hunk 18]
```diff
@@ -154,27 +168,66 @@ def __init__(
...
+
+        .. warning::
+
+            The default adjoint operator has small numerical errors due to interpolation. Set ``adjoint_via_backprop=True`` if you want to use the exact adjoint (computed via autograd).
+
+        :param torch.Tensor y: measurements
+        :return torch.Tensor: noisy measurements
+        """
+        if self.fan_beam or self.adjoint_via_backprop:
             assert (
```
Comment (Andrewwango) [inline deepinv/physics/tomography.py]: Change this to raise Value Error please

File: deepinv/physics/tomography.py [hunk 19]
```diff
@@ -125,6 +134,11 @@ def __init__(
         self.register_buffer("theta", theta)
 
         self.fan_beam = fan_beam
+        self.adjoint_via_backprop = adjoint_via_backprop
+        self.fbp_interpolate_boundary = fbp_interpolate_boundary
+        if circle:
+            # interpolate boundary does not make sense if circle is True
```
Comment (Andrewwango) [inline deepinv/physics/tomography.py:140]: Raise this as a warning?

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Just a couple of minor changes, apart from these it LGTM and happy to approve as soon as they are fixed! Thanks!
